File: money-manager-ios/VeralifyCore/Sources/VeralifyCore/MerchantKey.swift
import Foundation

/// The normalised merchant name that category rules, duplicate warnings and
/// reports key on.
///
/// Defined in docs/RECEIPTS_CONTRACTS.md §4 and implemented twice: here, and
/// in TypeScript in the AI gateway (`merchant-key.ts`), which files a receipt
/// using the rules this app writes. Both are tested against the same table
/// (`Fixtures/merchant_keys.json`), because a rule saved under one spelling
/// of the key and looked up under another is a correction the app silently
/// forgets.
public enum MerchantKey {

    /// Legal-form suffixes dropped from the end of the name, as token runs.
    static let legalSuffixes: [[String]] = [
        ["srl"], ["s", "r", "l"],
        ["spa"], ["s", "p", "a"],
        ["snc"], ["sas"],
        ["ltd"], ["limited"], ["plc"], ["llp"],
        ["inc"], ["llc"], ["gmbh"]
    ]

    /// `"ESSELUNGA S.p.A."` → `"esselunga"`, `"Marks & Spencer PLC"` → `"marks and spencer"`.
    ///
    /// Works on Unicode scalars and their general category rather than on
    /// `Character.isLetter`, so it agrees with the gateway's `\p{L}\p{N}` to
    /// the scalar. "Alphanumeric" includes non-Latin letters: an ASCII-only key
    /// would turn every Arabic merchant name into the same empty string.
    public static func make(from name: String) -> String {
        let decomposed = name.lowercased().decomposedStringWithCompatibilityMapping

        var tokens: [String] = []
        var current = String.UnicodeScalarView()
        func endToken() {
            if !current.isEmpty {
                tokens.append(String(current))
                current = String.UnicodeScalarView()
            }
        }

        for scalar in decomposed.unicodeScalars {
            if isMark(scalar) { continue }
            if scalar == "&" {
                endToken()
                tokens.append("and")
            } else if isAlphanumeric(scalar) {
                current.append(scalar)
            } else {
                endToken()
            }
        }
        endToken()

        // Repeated, because a name can carry two ("Foo Ltd Inc"), but never
        // down to nothing: a business literally called "SPA" keeps its name.
        var stripped = true
        while stripped {
            stripped = false
            for suffix in legalSuffixes
            where tokens.count > suffix.count && Array(tokens.suffix(suffix.count)) == suffix {
                tokens.removeLast(suffix.count)
                stripped = true
                break
            }
        }
        return tokens.joined(separator: " ")
    }

    private static func isMark(_ scalar: Unicode.Scalar) -> Bool {
        switch scalar.properties.generalCategory {
        case .nonspacingMark, .spacingMark, .enclosingMark: true
        default: false
        }
    }

    private static func isAlphanumeric(_ scalar: Unicode.Scalar) -> Bool {
        switch scalar.properties.generalCategory {
        case .uppercaseLetter, .lowercaseLetter, .titlecaseLetter, .modifierLetter, .otherLetter,
             .decimalNumber, .letterNumber, .otherNumber:
            true
        default:
            false
        }
    }
}

File: supabase/functions/ai-gateway/merchant-key.ts
// The normalised merchant name that rules, duplicate checks and reports key on.
//
// Defined in money-manager-ios/docs/RECEIPTS_CONTRACTS.md §4 and implemented
// twice: here, and in Swift as `MerchantKey` in VeralifyCore. Both are tested
// against tests/fixtures/merchant_keys.json, and a test here fails if the
// Swift copy of that table drifts, so a rule learned on the phone matches the
// receipt the gateway reads.

/** Legal-form suffixes dropped from the end, as token sequences. */
const LEGAL_SUFFIXES: readonly (readonly string[])[] = [
  ['srl'],
  ['s', 'r', 'l'],
  ['spa'],
  ['s', 'p', 'a'],
  ['snc'],
  ['sas'],
  ['ltd'],
  ['limited'],
  ['plc'],
  ['llp'],
  ['inc'],
  ['llc'],
  ['gmbh'],
];

/**
 * `"ESSELUNGA S.p.A."` → `"esselunga"`, `"Marks & Spencer PLC"` → `"marks and spencer"`.
 *
 * "Alphanumeric" means any Unicode letter or number, not just ASCII: the app
 * is used in Arabic too, and an ASCII-only key would collapse every Arabic
 * merchant name to the empty string — and so to the same merchant.
 */
export function merchantKey(name: string): string {
  const folded = name
    .toLowerCase()
    .normalize('NFKD')
    .replace(/\p{M}/gu, '')
    .replaceAll('&', ' and ');
  const tokens = folded.split(/[^\p{L}\p{N}]+/u).filter((t) => t.length > 0);

  // Repeated, because a name can carry two ("Foo Ltd Inc"), but never down to
  // nothing: a business literally called "SPA" keeps its name.
  let stripped = true;
  while (stripped) {
    stripped = false;
    for (const suffix of LEGAL_SUFFIXES) {
      if (tokens.length > suffix.length && endsWith(tokens, suffix)) {
        tokens.splice(tokens.length - suffix.length, suffix.length);
        stripped = true;
        break;
      }
    }
  }
  return tokens.join(' ');
}

function endsWith(tokens: string[], suffix: readonly string[]): boolean {
  const offset = tokens.length - suffix.length;
  return suffix.every((part, i) => tokens[offset + i] === part);
}
